File: android/app/src/main/java/sol/vortexx/android/files/api/FileTransferService.kt
package sol.vortexx.android.files.api

import kotlinx.coroutines.flow.Flow
import java.io.InputStream

/**
 * E2E-encrypted file up/downloads. Streams bytes so memory stays O(chunk)
 * even for gigabyte files. `Flow<TransferProgress>` lets the UI show a
 * progress bar without polling.
 */
interface FileTransferService {
    suspend fun upload(roomId: Long, source: InputStream, filename: String, sizeBytes: Long): Flow<TransferProgress>
    suspend fun download(roomId: Long, fileId: String, sink: (ByteArray) -> Unit): Flow<TransferProgress>
}

sealed interface TransferProgress {
    data class InFlight(val done: Long, val total: Long) : TransferProgress
    data class Done(val fileId: String, val ciphertextHashHex: String) : TransferProgress
    data class Error(val reason: String) : TransferProgress
}

File: ios/Modules/Sources/Files/api/FileTransferService.swift
import Foundation

/// E2E-encrypted file transfer. Streams chunks via `AsyncStream` so UI
/// can bind a progress bar without polling. Auto-switches to a chunked
/// resumable upload above a configurable threshold (Wave 14).
public protocol FileTransferService: Sendable {
    func upload(roomId: Int64, filename: String, data: Data) -> AsyncStream<TransferProgress>
    func download(roomId: Int64, fileId: String) -> AsyncStream<TransferProgress>
}

public enum TransferProgress: Sendable {
    case inFlight(done: Int64, total: Int64)
    case done(fileId: String, ciphertextHashHex: String, plaintext: Data?)
    case error(String)
}

public protocol MediaViewer: Sendable {
    func show(fileId: String, roomId: Int64, mimeType: String)
}
